### PR TITLE
Add Fixpoint and CoFixpoint in Search logical_kind

### PR DESCRIPTION
**Pull request by Villetaneuse**
This makes it possible e.g. to
```coq
Search (nat -> nat -> nat) [is:Fixpoint | is:Definition].
```

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  Here is [the updated `Search ` paragraph](https://coq.gitlabpages.inria.fr/-/coq/-/jobs/4333099/artifacts/_build/default/doc/refman-html/proof-engine/vernacular-commands.html?highlight=search#coq:cmd.Search)
  
  - [ ] Documented any new / changed **user messages**.
  - [x] Updated **documented syntax** by running `make doc_gram_rsts`.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by Villetaneuse** [comment]
I will wait before documenting, because I am not convinced it is the best solution.
For instance one could decide that `Fixpoint` and `CoFixpoint` should appear when `is:Definition` applies (the user should probably ignore what keyword was used to define some operation).

**Comment by herbelin** [comment]
> For instance one could decide that `Fixpoint` and `CoFixpoint` should appear when `is:Definition` applies

The current idea with `is:` is really that it refers to the word used in the source code. But, indeed, that may be more intuitive to decide that `is:Definition` covers, say, everything that is transparent.

If you want to explore this alternative approach (in addition to providing `is:Fixpoint` and `is:CoFixpoint`), I suspect that refining [current line 225](https://github.com/coq/coq/blob/5e0a71c3bdbd097f14f2a8a4e4377cd1ec03ab0d/vernac/search.ml#L225) of `search.ml` into say:
```ocaml
| GlobSearchKind k -> (match kind with None -> false | Some k' -> included_kind k' k)
```
together with say:
```ocaml
let included_kind k' k =
  match k', k with
  | IsDefinition _, IsDefinition Definition -> true
  | _ -> k = k'
```
would do the job.

**Comment by Villetaneuse** [comment]
> The current idea with `is:` is really that it refers to the word used in the source code. But, indeed, that may be more intuitive to decide that `is:Definition` covers, say, everything that is transparent.

Then let's keep it simple and consistent with `is:Lemma`, etc. Then, a competent contributor (which could be a future me), can still add, e.g. `kind:Transparent` and `kind:Opaque`, or something like that.

Do you think it's worth writing a small test?

**Comment by herbelin** [comment]
> Do you think it's worth writing a small test?

I would say yes and you can extend one of the existing `Search`something file of the test-suite.

**Comment by Villetaneuse** [comment]
Also looks like, for some reason, my `rst` is again wrong... Please, someone, help me on this, `rst` is the straw that broke the camel's (no pun intented) back with respect to the zillion languages/build systems/... needed for Coq contributing.

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 1]
```diff
@@ -188,15 +188,17 @@ described elsewhere
       .. insertprodn logical_kind logical_kind
 
       .. prodn::
-         logical_kind ::= {| @thm_token | @assumption_token }
+         logical_kind ::=
+         | {| @thm_token | @assumption_token }
+         | {| Fixpoint | CoFixpoint }
```
Comment (SkySkimmer) [inline doc/sphinx/proof-engine/vernacular-commands.rst:193]: don't update this stuff by hand, reset it then `make doc_gram_rsts`

**Comment by JasonGross** [comment]
> The current idea with `is:` is really that it refers to the word used in the source code. But, indeed, that may be more intuitive to decide that `is:Definition` covers, say, everything that is transparent.

> Then let's keep it simple and consistent with `is:Lemma`, etc. Then, a competent contributor (which could be a future me), can still add, e.g. `kind:Transparent` and `kind:Opaque`, or something like that.

Maybe we should go with `is:Defined` and `is:Qed`?  `kind:Transparent` and `kind:Opaque` might be too confusing given that `Opaque` and `Transparent` exist as vernaculars.  Or `has:TransparentBody` and `has:OpaqueBody` (and `is:Axiom`)?

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 1 above)
Comment (Villetaneuse) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: Thanks! Done (it doesn't look that good though).

**Comment by Villetaneuse** [comment]
> > The current idea with `is:` is really that it refers to the word used in the source code. But, indeed, that may be more intuitive to decide that `is:Definition` covers, say, everything that is transparent.
> 
> > Then let's keep it simple and consistent with `is:Lemma`, etc. Then, a competent contributor (which could be a future me), can still add, e.g. `kind:Transparent` and `kind:Opaque`, or something like that.
> 
> Maybe we should go with `is:Defined` and `is:Qed`? `kind:Transparent` and `kind:Opaque` might be too confusing given that `Opaque` and `Transparent` exist as vernaculars. Or `has:TransparentBody` and `has:OpaqueBody` (and `is:Axiom`)?

Thank you for your suggestion. Anything which is not too verbose (we want to search quickly) and explicit would be fine for me. Maybe we can open an issue in order not to forget?

File: doc/tools/docgram/orderedGrammar [hunk 2]
```diff
@@ -1053,6 +1053,8 @@ search_item: [
 ]
 
 logical_kind: [
+| "Fixpoint"
+| "CoFixpoint"
 | [ thm_token | assumption_token ]
 | [ "Definition" | "Example" | "Context" | "Primitive" | "Symbol" ]
 | [ "Coercion" | "Instance" | "Scheme" | "Canonical" | "SubClass" ]
```
Comment (SkySkimmer) [inline doc/tools/docgram/orderedGrammar:1060]: editing this to be
```suggestion
| [ "Coercion" | "Instance" | "Scheme" | "Canonical" | "SubClass" | "Fixpoint" | "CoFixpoint" ]
```
(or whatever order you prefer) 
may work

Comment (Villetaneuse) [reply doc/tools/docgram/orderedGrammar:1060]: Hmm, didn't work. Maybe I did not do what was needed to please the gods. I would say it's not worth fighting for.

Comment (Villetaneuse) [reply doc/tools/docgram/orderedGrammar:1060]: @coqbot run full ci

**Comment by Villetaneuse** [comment]
@coqbot run full ci

**Comment by Villetaneuse** [comment]
> > The current idea with `is:` is really that it refers to the word used in the source code. But, indeed, that may be more intuitive to decide that `is:Definition` covers, say, everything that is transparent.
> 
> > Then let's keep it simple and consistent with `is:Lemma`, etc. Then, a competent contributor (which could be a future me), can still add, e.g. `kind:Transparent` and `kind:Opaque`, or something like that.
> 
> Maybe we should go with `is:Defined` and `is:Qed`? `kind:Transparent` and `kind:Opaque` might be too confusing given that `Opaque` and `Transparent` exist as vernaculars. Or `has:TransparentBody` and `has:OpaqueBody` (and `is:Axiom`)?

I have filled [this issue](https://github.com/coq/coq/issues/18985).

File: doc/tools/docgram/orderedGrammar (hunk 2 above)
Comment (SkySkimmer) [reply doc/tools/docgram/orderedGrammar]: I pushed a commit

Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar]: @Villetaneuse The grammar documentation tool is described in `doc/tools/docgram/README.md`

File: doc/tools/docgram/common.edit_mlg [hunk 3]
```diff
@@ -1971,7 +1971,9 @@ logical_kind: [
 | DELETE "Instance"
 | DELETE "Scheme"
 | DELETE "Canonical"
-| [  "Coercion" | "Instance" | "Scheme" | "Canonical" | "SubClass" ]
+| DELETE "Fixpoint"
+| DELETE "CoFixpoint"
+| [  "Coercion" | "Instance" | "Scheme" | "Canonical" | "SubClass" | "Fixpoint" | "CoFixpoint" ]
 | DELETE "Field"
 | DELETE "Method"
 | [ "Field" | "Method" ]
```
[the remark this answers is not in the document]
Comment (Villetaneuse) [reply doc/tools/docgram/common.edit_mlg]: Thank you very much @jfehrle!

**Comment by Villetaneuse** [comment]
@coqbot run full ci

**Comment by Villetaneuse** [comment]
The refman looks good thanks to @jfehrle. This PR is probably ready do be merged.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**